src/core/algorithms/nar/des/encoded_value_range.h

```diff
@@ -0,0 +1,112 @@
+#pragma once
+
+#include "algorithms/nar/value_range.h"
+#include "model/table/column_layout_typed_relation_data.h"
+#include "rng.h"
+
+namespace algos::des {
+
+class EncodedValueRange {
+public:
+    static size_t const kFieldCount = 4;
+    double permutation;
+    double threshold;
+    double bound1;
+    double bound2;
+
+    // TODO: remove code duplication here
+    double& operator[](size_t index) {
+        switch (index) {
+            case 0:
+                return permutation;
+                break;
+            case 1:
+                return threshold;
+                break;
+            case 2:
+                return bound1;
+                break;
+            case 3:
+                return bound2;
+                break;
+            default:
+                throw std::out_of_range("Index out of range for value range.");
+        }
+    }
+
+    double const& operator[](size_t index) const {
+        switch (index) {
+            case 0:
+                return permutation;
+                break;
+            case 1:
+                return threshold;
+                break;
+            case 2:
+                return bound1;
+                break;
+            case 3:
+                return bound2;
+                break;
+            default:
+                throw std::out_of_range("Index out of range for value range.");
+        }
+    }
+
+    std::shared_ptr<model::ValueRange> Decode(std::shared_ptr<model::ValueRange> domain) const {
+        switch (domain->GetTypeId()) {
+            case model::TypeId::kInt: {
+                std::shared_ptr<model::IntValueRange> int_domain =
+                        std::static_pointer_cast<model::IntValueRange>(domain);
+                model::Int span = int_domain->upper_bound - int_domain->lower_bound;
+                model::Int resulting_lower = int_domain->lower_bound + span * this->bound1;
+                model::Int resulting_upper = int_domain->lower_bound + span * this->bound2;
+                if (resulting_lower > resulting_upper) {
+                    std::swap(resulting_lower, resulting_upper);
+                }
+                return std::make_shared<model::IntValueRange>(
+                        model::IntValueRange(resulting_lower, resulting_upper));
```

Comment:
```suggestion
                return std::make_shared<model::IntValueRange>(resulting_lower, resulting_upper);
```
The purpose of `make_shared` is to forward it's parameters to `IntValueRange`'s constructor and create `IntValueRange` in-place. In your code temporary `IntValueRange` is created and being copied (or moved) to `shared_ptr`. Also it's not a good idea to repeat type name.
This also applies to `Double` and `Mixed` cases